src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -203,69 +251,241 @@ public object[] ArgumentList
+
+        #region PSTasks
+
+        private PSTaskPool _taskPool;
+        private PSTaskDataStreamWriter _taskDataStreamWriter;
+        private Dictionary<string, object> _usingValuesMap;
+        private System.Threading.Timer _taskTimer;
+        private PSTaskJob _taskJob;
+
+        private void InitParallelParameterSet()
+        {
+            bool allowUsingExpression = this.Context.SessionState.LanguageMode != PSLanguageMode.NoLanguage;
+            _usingValuesMap = ScriptBlockToPowerShellConverter.GetUsingValuesAsDictionary(Parallel, allowUsingExpression, this.Context, null);
+            
+            // Validate using values map
+            foreach (var item in _usingValuesMap.Values)
+            {
+                if (item is ScriptBlock)
                 {
-                    whatIf = (SwitchParameter)argument;
+                    this.ThrowTerminatingError(
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelUsingVariableCannotBeScriptBlock),
+                            "ParallelUsingVariableCannotBeScriptBlock",
+                            ErrorCategory.InvalidType,
+                            this));
                 }
+            }
 
-                if (psBoundParameters.TryGetValue("confirm", out argument))
+            if (AsJob)
+            {
+                if (MyInvocation.BoundParameters.ContainsKey(nameof(TimeoutSeconds)))
                 {
-                    confirm = (SwitchParameter)argument;
+                    this.ThrowTerminatingError(
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelCannotUseTimeoutWithJob),
+                            "ParallelCannotUseTimeoutWithJob",
+                            ErrorCategory.InvalidOperation,
+                            this));
                 }
 
-                if (whatIf || confirm)
+                _taskJob = new PSTaskJob(
+                    Parallel.ToString(),
+                    ThrottleLimit);
+            }
+            else
+            {
+                _taskDataStreamWriter = new PSTaskDataStreamWriter(this);
+                _taskPool = new PSTaskPool(ThrottleLimit);
+                _taskPool.PoolComplete += (sender, args) =>
                 {
-                    string message = InternalCommandStrings.NoShouldProcessForScriptBlockSet;
-
-                    ErrorRecord errorRecord = new ErrorRecord(
-                        new InvalidOperationException(message),
-                        "NoShouldProcessForScriptBlockSet",
-                        ErrorCategory.InvalidOperation,
-                        null);
-                    ThrowTerminatingError(errorRecord);
+                    _taskDataStreamWriter.Close();
+                };
+                if (TimeoutSeconds != 0)
+                {
+                    _taskTimer = new System.Threading.Timer(
```

Comment:
Is it possible to have `using System.Threading;` at the top of the file to obviate this namespace qualification?